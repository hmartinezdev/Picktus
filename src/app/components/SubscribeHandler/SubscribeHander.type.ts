import { ThunkResult } from '@store/reducers/user/user-actions';

export interface IStringMap {
  [s: string]: string;
}
export interface ISubscribeHandlerState {
  open: boolean;
  inputs: IStringMap;
  errors: IStringMap;
}

export interface IReduxStateProps {
  loading?: boolean;
}

export interface IDispatchProps {
  userCreation: (mail: string, password: string) => ThunkResult<void>;
}

export type ISubscribeHandlerProps = IReduxStateProps & IDispatchProps;
