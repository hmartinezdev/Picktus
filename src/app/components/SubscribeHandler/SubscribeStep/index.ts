import SubscribeStep from './SubscribeStep';

export default SubscribeStep;
