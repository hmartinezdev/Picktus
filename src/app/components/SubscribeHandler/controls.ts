/**
 * Detect if string is a valid email address
 *
 * @param  {string} mail
 * @returns boolean
 */
export const isMail = (mail: string): boolean => {
  const mailRegex = new RegExp( // tslint:disable-next-line
    /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
  );
  return !!mail.match(mailRegex);
};
/**
 * Detect if the string is a strong enough password
 *
 * @param  {string} password
 * @returns boolean
 */
export const isPasswordSecure = (password: string): boolean => {
  const passwordRegex = new RegExp('^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*-]).{8,}$');

  return !!password.match(passwordRegex);
};
