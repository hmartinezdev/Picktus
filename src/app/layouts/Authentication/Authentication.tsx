import Logo from '@components/Logo';
import colors from '@constants/colors';
import { borderRadius, boxShadow } from '@constants/styles';
import React, { Component } from 'react';
import { Transition, TransitionGroup } from 'react-transition-group';

export interface IAuthenticationPropsType {
  status: string;
  children: JSX.Element;
}

class Authentication extends Component<IAuthenticationPropsType> {
  public render(): React.ReactElement<Authentication> {
    const { status, children } = this.props;

    return (
      <div className={`container container--${status}`}>
        <div className={`logo`}>
          <Logo />
        </div>
        <TransitionGroup className="authentication__content" appear={true}>
          {React.Children.map(children, (child: JSX.Element) => (
            <Transition key={child.key || ''} timeout={300} mountOnEnter={true} unmountOnExit={true} appear={true}>
              {(contentStatus) => <div className={`contentContainer contentContainer--${contentStatus}`}>{child}</div>}
            </Transition>
          ))}
        </TransitionGroup>
        <style jsx>
          {`
            .container {
              width: 100%;
              height: 100vh;
              display: flex;
              align-items: center;
              flex-direction: column;
              padding: 2rem 0;
              box-sizing: border-box;
              overflow: hidden;
              opacity: 1;
              transition: opacity 300ms ease-out;
            }

            .container--exiting,
            .container--exited {
              position: absolute;
              z-index: 0;
            }

            .contentContainer {
              transition: opacity 300ms ease-out;
            }

            .contentContainer--entering {
              animation: opacity 300ms ease-in;
            }

            .contentContainer--entered {
              opacity: 1;
            }

            .contentContainer--exiting,
            .contentContainer--exited {
              position: absolute;
              opacity: 0;
            }

            .logo {
              background-color: ${colors.primary};
              box-shadow: ${boxShadow};
              display: flex;
              position: relative;
              border-radius: ${borderRadius};
              transition: opacity 300ms ease-out;
            }

            :global(.authentication__content) {
              height: 100%;
              width: 100%;
              display: flex;
              flex-direction: column;
              justify-content: flex-start;
              align-items: center;
              margin-top: 2.5rem;
            }

            @keyframes opacity {
              0% {
                opacity: 0;
              }

              100% {
                opacity: 1;
              }
            }
          `}
        </style>
      </div>
    );
  }
}

export default Authentication;
