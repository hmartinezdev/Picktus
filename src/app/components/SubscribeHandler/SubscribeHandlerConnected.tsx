import { IReduxState } from '@store/reducers';
import { userCreation } from '@store/reducers/user/user-actions';
import { connect } from 'react-redux';
import { IReduxStateProps } from './SubscribeHander.type';
import SubscribeHandler from './SubscribeHandler';

export const mapStateToProps = (state: IReduxState): IReduxStateProps => {
  return {
    loading: !!state.user.creationInProgress,
  };
};

export default connect(
  mapStateToProps,
  { userCreation }
)(SubscribeHandler);
