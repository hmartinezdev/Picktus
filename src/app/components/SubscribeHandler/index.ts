import SubscribeHandler from './SubscribeHandler';

export default SubscribeHandler;
