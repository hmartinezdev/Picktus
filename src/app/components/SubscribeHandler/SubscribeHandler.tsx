import FormPagination from '@components/FormPagination/FormPagination';
import Loader from '@components/Loader';
import colors from '@constants/colors';
import { borderRadius, boxShadow } from '@constants/styles';
import React, { PureComponent } from 'react';
import { Transition, TransitionGroup } from 'react-transition-group';
import { isMail, isPasswordSecure } from './controls';
import { ISubscribeHandlerProps, ISubscribeHandlerState, ISubscribeStepInfos } from './SubscribeHander.type';
import SubscribeStep from './SubscribeStep';

class SubscribeHandler extends PureComponent<ISubscribeHandlerProps, ISubscribeHandlerState> {
  public static defaultProps = {
    requestStatus: null,
  };

  public form: ISubscribeStepInfos[] = [];

  constructor(props: ISubscribeHandlerProps) {
    super(props);
    this.state = { current: 0, values: {} };

    this.form = [
      {
        control: isMail,
        errorMessage: 'You need to user a valid email address',
        name: 'mail',
        title: 'Email address',
        type: 'text',
      },
      {
        control: isPasswordSecure,
        errorMessage:
          'Password must be minimum eight characters,  at least one letter, one number and one special character',
        name: 'password',
        title: 'Password',
        type: 'password',
      },
      {
        control: this.confirmPassword,
        errorMessage: 'The password confirmation must be identical to the password',
        name: 'confirm',
        title: 'Password confirmation',
        type: 'password',
      },
    ];
  }

  public componentWillReceiveProps(nextProps: ISubscribeHandlerProps) {
    const { requestStatus } = this.props;
    const nextRequestStatus = nextProps.requestStatus;
    if (requestStatus && nextRequestStatus) {
      if (requestStatus.inProgress && !nextRequestStatus.inProgress) {
        this.setState({ current: 0 });
      }
    }
  }

  public confirmPassword = (passwordConfirmation: string): boolean =>
    passwordConfirmation === this.state.values.password;

  public onValidate = (value: string): void => {
    if (this.state.current === this.form.length - 1) {
      this.onLastValueSubmitted();
    }

    this.setState({
      current: this.state.current + 1,
      values: { ...this.state.values, [this.form[this.state.current].name]: value },
    });
  };

  public onLastValueSubmitted = (): void => {
    const { userCreation } = this.props;

    userCreation(this.state.values.mail, this.state.values.password);
  };

  public onValidatedStepClick = (step: number): void => {
    this.setState({ current: step });
  };

  public render(): React.ReactElement<SubscribeHandler> {
    return (
      <div className="container">
        <FormPagination
          current={this.state.current}
          steps={this.form.reduce((accumulator: string[], value) => [...accumulator, value.title], [])}
          onValidatedStepClick={this.onValidatedStepClick}
        />
        <div className="form">
          <TransitionGroup component={null}>
            {this.state.current <= this.form.length - 1 ? (
              this.form.filter((value, index) => index === this.state.current).map((value) => (
                <Transition timeout={300} mountOnEnter unmountOnExit in appear>
                  {(status) => (
                    <div key={value.name} className={`stepContainer stepContainer--${status}`}>
                      <SubscribeStep
                        onValidate={this.onValidate}
                        control={value.control}
                        errorMessage={value.errorMessage}
                        name={value.name}
                        title={value.title}
                        type={value.type}
                      />
                    </div>
                  )}
                </Transition>
              ))[0]
            ) : (
              <Loader />
            )}
          </TransitionGroup>
        </div>
        <style jsx>{`
          .container {
            box-shadow: ${boxShadow};
            background-color: ${colors.primary};
            width: 24rem;
            padding: 0.7rem;
            box-sizing: border-box;
            border-radius: ${borderRadius};
            display: flex;
            align-items: center;
            flex-direction: column;
          }

          .stepContainer {
            opacity: 0;
            transition: opacity 300ms ease-in;
            width: 100%;
          }

          .stepContainer--entered,
          .stepContainer--entering {
            opacity: 1;
          }

          .form {
            display: flex;
            flex-direction: column;
            align-items: center;
            width: 16rem;
            height: 5rem;
          }

          .inputTransition {
            width: 100%;
          }
        `}</style>
      </div>
    );
  }
}

export default SubscribeHandler;
