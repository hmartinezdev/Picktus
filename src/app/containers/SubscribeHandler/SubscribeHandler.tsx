import ButtonLink from '@components/ButtonLink';
import FormPagination from '@components/FormPagination/FormPagination';
import Loader from '@components/Loader';
import colors from '@constants/colors';
import { borderRadius, boxShadow } from '@constants/styles';
import React, { PureComponent } from 'react';
import { Transition, TransitionGroup } from 'react-transition-group';
import { isMail, isPasswordSecure } from './controls';
import { ISubscribeHandlerProps, ISubscribeHandlerState, ISubscribeStepInfos } from './SubscribeHandler.type';
import SubscribeStepConnected from './SubscribeStep';

class SubscribeHandler extends PureComponent<ISubscribeHandlerProps, ISubscribeHandlerState> {
  public static defaultProps = {
    requestStatus: null,
  };

  public form: ISubscribeStepInfos[] = [];

  constructor(props: ISubscribeHandlerProps) {
    super(props);
    this.state = { current: 0, values: {} };

    this.form = [
      {
        control: isMail,
        errorMessage: 'You need to user a valid email address',
        name: 'mail',
        title: 'Email address',
        type: 'text',
      },
      {
        control: isPasswordSecure,
        errorMessage:
          'Password must be minimum eight characters,  at least one letter, one number and one special character',
        name: 'password',
        title: 'Password',
        type: 'password',
      },
      {
        control: this.confirmPassword,
        errorMessage: 'The password confirmation must be identical to the password',
        name: 'confirm',
        title: 'Password confirmation',
        type: 'password',
      },
    ];
  }

  public componentWillReceiveProps(nextProps: ISubscribeHandlerProps) {
    const { requestStatus } = this.props;
    const nextRequestStatus = nextProps.requestStatus;
    if (requestStatus && nextRequestStatus) {
      if (requestStatus.inProgress && !nextRequestStatus.inProgress) {
        this.setState({ current: 0 });
      }
    }
  }

  public confirmPassword = (passwordConfirmation: string): boolean =>
    passwordConfirmation === this.state.values.password;

  public onValidate = (value: string): void => {
    if (this.state.current === this.form.length - 1) {
      this.onLastValueSubmitted();
    }

    this.setState({
      current: this.state.current + 1,
      values: { ...this.state.values, [this.form[this.state.current].name]: value },
    });
  };

  public onLastValueSubmitted = (): void => {
    const { userCreation } = this.props;

    userCreation(this.state.values.mail, this.state.values.password);
  };

  public onValidatedStepClick = (step: number): void => {
    this.setState({ current: step });
  };

  public onPreviousClick = () => {
    const { current } = this.state;
    if (current > 0) {
      this.setState({ current: current - 1 });
    }
  };

  public render(): React.ReactElement<SubscribeHandler> {
    const currentStep = this.form[this.state.current];
    return (
      <div className="container">
        <div className="formContainer">
          {' '}
          <FormPagination
            current={this.state.current}
            steps={this.form.reduce((accumulator: string[], value) => [...accumulator, value.title], [])}
            onValidatedStepClick={this.onValidatedStepClick}
          />
          <div className="form">
            <TransitionGroup component={null}>
              {this.state.current <= this.form.length - 1 ? (
                this.form
                  .filter((_value, index) => index === this.state.current)
                  .map((value) => (
                    <Transition timeout={300} key={`${value.name}-container`} mountOnEnter unmountOnExit in appear>
                      {(status: string) => (
                        <div key={value.name} className={`stepContainer stepContainer--${status}`}>
                          <SubscribeStepConnected
                            onValidate={this.onValidate}
                            control={value.control}
                            errorMessage={value.errorMessage}
                            name={value.name}
                            title={value.title}
                            type={value.type}
                            defaultValue={currentStep ? this.state.values[currentStep.name] : ''}
                            current={this.state.current}
                            onPreviousClick={this.onPreviousClick}
                          />
                        </div>
                      )}
                    </Transition>
                  ))[0]
              ) : (
                <Transition timeout={300} key={'loader-container'} mountOnEnter unmountOnExit in appear>
                  {(status: string) => (
                    <div key="loader" className={`stepContainer stepContainer--${status}`}>
                      <Loader />
                    </div>
                  )}
                </Transition>
              )}
            </TransitionGroup>
          </div>
        </div>

        <ButtonLink prefetch={true} text="I have an account!" href="/auth/login" dark />
        <style jsx>{`
          .formContainer {
            box-shadow: ${boxShadow};
            background-color: ${colors.primary};
            width: 24rem;
            padding: 0.7rem;
            box-sizing: border-box;
            border-radius: ${borderRadius};
            display: flex;
            align-items: center;
            flex-direction: column;
            height: 16rem;
          }

          .stepContainer {
            opacity: 0;
            transition: opacity 300ms ease-in;
            width: 100%;
            height: 100%;
          }

          .stepContainer--entered,
          .stepContainer--entering {
            opacity: 1;
          }

          .stepContainer--exiting,
          .stepContainer--exited {
            position: absolute;
          }

          .form {
            display: flex;
            flex-direction: column;
            align-items: center;
            width: 16rem;
            position: relative;
          }

          .inputTransition {
            width: 100%;
          }
        `}</style>
      </div>
    );
  }
}

export default SubscribeHandler;
