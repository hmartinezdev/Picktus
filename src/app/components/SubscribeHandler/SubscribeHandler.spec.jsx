import Authentication from '@services/authentication';
import { mount, shallow } from 'enzyme';
import React from 'react';
import SubscribeHandler from './SubscribeHandler';

let wrapper;
let instance;
let mountWrapper;

const baseProps = {
  requestStatus: {
    inProgress: false,
    error: 'error',
  },
};

const setup = (props = {}) => shallow(<SubscribeHandler {...baseProps} {...props} />);

describe('<SubscribeHandler />', () => {
  beforeEach(() => {
    wrapper = setup();
    instance = wrapper.instance();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('componentWillReceiveProps', () => {
    test('it should call setState with current at 0 if the user creation progress has finished', () => {
      wrapper = setup({ requestStatus: { inProgress: true } });
      const spy = jest.spyOn(wrapper, 'setState');
      wrapper.setProps({ requestStatus: { inProgress: false } });
      expect(spy).toHaveBeenCalledWith({ current: 0 });
    });
  });

  describe('confirmPassWord', () => {
    test('it should return true if the password value is equal to the confirmation', () => {
      wrapper.setState({ values: { password: 'test' } });
      instance = wrapper.instance();
      expect(instance.confirmPassword('test')).toBe(true);
    });

    test('it should return false if not', () => {
      wrapper.setState({ values: { password: 'test' } });
      instance = wrapper.instance();
      expect(instance.confirmPassword('test1')).toBe(false);
    });
  });

  describe('onValidate', () => {
    test('it should call onLastValueSubmitted if current is equal to the number of steps minus 1', () => {
      wrapper = setup({ userCreation: () => undefined });
      instance = wrapper.instance();
      instance.setState({ current: instance.form.length - 1 });
      const spy = jest.spyOn(instance, 'onLastValueSubmitted');
      instance.onValidate();
      expect(spy).toHaveBeenCalled();
    });

    test("it should increment current value and store the value in the component state if it's not", () => {
      const spy = jest.spyOn(instance, 'setState');
      instance.onValidate('test');
      expect(spy).toBeCalledWith({ current: 1, values: { [instance.form[0].name]: 'test' } });
    });
  });

  describe('onLastValueSubmitted', () => {
    test('it should call userCreation with the defined password and email', () => {
      const userCreation = jest.fn();
      wrapper = setup({ userCreation });
      wrapper.setState({ values: { password: 'password', mail: 'mail' } });
      instance = wrapper.instance();
      instance.onLastValueSubmitted();
      expect(userCreation).toHaveBeenCalledWith('mail', 'password');
    });
  });

  describe('onValidatedStepClick', () => {
    test('it should set the state to the clicked step', () => {
      const spy = jest.spyOn(instance, 'setState');
      instance.onValidatedStepClick(1);
      expect(spy).toHaveBeenCalledWith({ current: 1 });
    });
  });
});
