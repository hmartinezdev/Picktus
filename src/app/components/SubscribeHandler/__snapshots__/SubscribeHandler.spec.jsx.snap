// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubscribeHandler /> should render properly 1`] = `
<div
  className="container"
>
  <FormPagination
    current={0}
    onValidatedStepClick={[Function]}
    steps={
      Array [
        "Email address",
        "Password",
        "Password confirmation",
      ]
    }
  />
  <div
    className="form"
  >
    <TransitionGroup
      childFactory={[Function]}
      component={null}
    >
      <Transition
        appear={true}
        enter={true}
        exit={true}
        in={true}
        mountOnEnter={true}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={300}
        unmountOnExit={true}
      >
        <Component />
      </Transition>
    </TransitionGroup>
  </div>
  <style
    jsx={true}
  >
    
          .container {
            box-shadow: 11px 10px 34px -7px rgba(0, 0, 0, 0.49);
            background-color: #2d3047;
            width: 24rem;
            padding: 0.7rem;
            box-sizing: border-box;
            border-radius: 3px;
            display: flex;
            align-items: center;
            flex-direction: column;
          }

          .stepContainer {
            opacity: 0;
            transition: opacity 300ms ease-in;
            width: 100%;
          }

          .stepContainer--entered,
          .stepContainer--entering {
            opacity: 1;
          }

          .form {
            display: flex;
            flex-direction: column;
            align-items: center;
            width: 16rem;
            height: 5rem;
          }

          .inputTransition {
            width: 100%;
          }
        
  </style>
</div>
`;
